anki_addon/sonaveeb.py

```diff
@@ -30,44 +61,70 @@ class WordInfo:
     word_id: int = None
     word: str = None
     word_class: str = None
+    word_audio_url: str = None
     url: str = None
     lexemes: tp.List[LexemeInfo] = None
-    morphology: tp.List[tp.Tuple[str]] = None
-    audio_urls: tp.List[str] = dc.field(default_factory=list)
-
-    def summary(self, lang=None):
-        data = {
-            'word_id': self.word_id,
-            'word': self.word,
-            'word_class': self.word_class,
-            'url': self.url,
-            'short_record': self.short_record(),
-            'morphology': self.morphology,
-            'lexemes': self.lexemes,
-        }
-        result = '\n'.join([f'{k}: {v}' for k, v in data.items() if v is not None])
-        return result
-
-    def short_record(self):
-        forms = [form[0] for form in self.morphology if len(form) > 0]
-        if len(forms) > 2:
-            p1 = os.path.commonprefix([forms[0], forms[1]])
-            p2 = os.path.commonprefix([forms[0], forms[1]])
-            prefix = p1 if len(p1) > len(p2) else p2
-            if len(prefix) > 3 or prefix == forms[0]:
-                if forms[0] == prefix:
-                    short = forms[0]
-                else:
-                    short = forms[0].replace(prefix, f'{prefix}/')
-                for form in forms[1:]:
-                    short += ', ' + form.replace(prefix, '-')
-            else:
-                short = ', '.join(forms)
-        elif len(forms) > 0:
-            short = forms[0]
+    morphology: tp.Dict[str, str] = dc.field(default_factory=dict)
+    morphology_audio_urls: tp.Dict[str, str] = dc.field(default_factory=dict)
+
+    def __post_init__(self):
+        # Validate present form types
+        if missing_forms := self.missing_form_types():
+            logging.error(
+                'Missing expected forms for the word '
+                f'"{self.word}" [##{self.word_id}]: {missing_forms}'
+            )
+
+    def audio_urls(self) -> tp.List[str]:
+        '''Returns a list audio URLs corresponding to essential_forms list.
+
+        Not every form is guaranteed to have an audio, but every audio is
+        guaranteed to be one of the essential forms.
+        '''
+        keys = self.present_form_types()
+        urls = [self.morphology_audio_urls.get(key) for key in keys]
+        urls = [u for u in urls if u is not None]
+        if not urls and self.word_audio_url is not None:
+            urls = [self.word_audio_url]
+        return urls
+
+    def essential_forms(self, compress=False, join=False) -> tp.List[str] | str:
```

Comment:
This is incompatible with Python 3.9, which is used in Anki for macOS.

```suggestion
    def essential_forms(self, compress=False, join=False) -> tp.Union[tp.List[str], str]:
```

Reply:
Thanks!